src/d3a/d3a_core/simulation.py

```diff
@@ -255,20 +254,21 @@ def _update_and_send_results(self):
         self.endpoint_buffer.update_stats(
             self.area, self.status, self.progress_info, self.current_state)
         self.update_area_stats(self.area, self.endpoint_buffer)
-        if self.export_on_finish and self.should_export_results and \
+        if self.export_results_on_finish and \
                 self.area.current_market is not None and d3a.constants.D3A_TEST_RUN:
+            # for integration tests:
             self.export.raw_data_to_json(
                 self.area.current_market.time_slot_str,
                 self.endpoint_buffer.flattened_area_core_stats_dict
             )
-        if self.should_export_results:
+        if self.export_results_on_finish:
             self.file_stats_endpoint(self.area)
-            return
 
-        results = self.endpoint_buffer.prepare_results_for_publish()
-        if results is None:
-            return
-        self.kafka_connection.publish(results, self._simulation_id)
+        if self.should_send_results:
+            results = self.endpoint_buffer.prepare_results_for_publish()
+            if results is None:
+                return
+            self.kafka_connection.publish(results, self._simulation_id)
```

Comment:
I would further refactor all this to the following structure:
`if self.export_results_on_finish: .... `
`elif self.should_send_results: ....`
Reason for that is because these statements are mutually exclusive, therefore we are certain that if we are exporting the simulation results we should not send them via the message broker.